src/app/shared/services/doctors.service.ts

```diff
@@ -1,32 +1,160 @@
 import { AngularFireDatabase } from '@angular/fire/compat/database';
+import { BehaviorSubject, Subscription } from 'rxjs';
+
+import { Doctor } from '../components/doctor-card/doctor.model';
+import { SortingTypes } from '../models/sorting-types.model';
+import { DoctorsSpecialty } from '../models/doctors-specialty.model';
+import { get } from 'lodash';
+
+const DOCTORS_PER_PAGE = 6;
 
 @Injectable({
   providedIn: 'root',
 })
-export class DoctorsService {
-  doctorsSubject = new BehaviorSubject<Doctor[]>([]);
-  doctors = this.doctorsSubject.asObservable();
+export class DoctorsService implements OnDestroy {
+  private visibleDoctorsSubject = new BehaviorSubject<Doctor[]>([]);
+  private visibleSpecialtiesSubject = new BehaviorSubject<DoctorsSpecialty[]>(
+    []
+  );
+
+  private topDoctorsSubject = new BehaviorSubject<Doctor[]>([]);
+  private paginationSizeSubject = new BehaviorSubject<number>(0);
+
+  private paginationCurrentSubject = new BehaviorSubject<number>(0);
+  private sortingTypeSubject = new BehaviorSubject<string>('');
+  private searchQuerySubject = new BehaviorSubject<string>('');
+
+  private subsription!: Subscription;
+
+  visibleDoctors = this.visibleDoctorsSubject.asObservable();
+  visibleSpecialties = this.visibleSpecialtiesSubject.asObservable();
+
+  paginationCurrent = this.paginationCurrentSubject.asObservable();
+  paginationSize = this.paginationSizeSubject.asObservable();
+  sortingType = this.sortingTypeSubject.asObservable();
+
+  topDoctors = this.topDoctorsSubject.asObservable();
+
+  private firebaseDoctors: Doctor[];
+  private firebaseSpecialties: Object;
 
   constructor(private db: AngularFireDatabase) {
-    this.doctors = this.db.list('/doctors').valueChanges() as Observable<
-      Doctor[]
-    >;
+    this.firebaseDoctors = [];
+    this.firebaseSpecialties = {};
   }
 
-  getTopDoctors(): Observable<Doctor[]> {
-    return this.doctors.pipe(
-      map((doctors) => doctors.filter((doctor) => doctor.top))
-    );
+  ngOnDestroy(): void {
+    this.subsription.unsubscribe();
+  }
+
+  loadData(): void {
+    this.subsription = this.db
+      .list('/doctors')
+      .valueChanges()
+      .subscribe((data) => {
+        this.firebaseDoctors = data as Doctor[];
+        this.updateVisibleDoctors();
+        this.calculateTopDoctors();
+      });
+
+    this.loadSpecialties();
   }
 
-  getSpecialties(): Observable<string[]> {
-    return this.doctors.pipe(
-      map((doctors) =>
-        Array.from(new Set(doctors.map((doctor) => doctor.specialization)))
-      )
+  loadSpecialties(): void {
+    this.db.database.ref('specialties').once('value', (snap) => {
+      this.firebaseSpecialties = snap.val();
+      const specialties = Object.entries(snap.val()).map(([id, name]) => ({
+        id,
+        name: name as string,
+        checked: false,
+      }));
+      this.visibleSpecialtiesSubject.next(specialties);
+    });
+  }
+
+  private updateVisibleDoctors(): void {
```

Comment:
Suggest separation of concerns and move all filtering logic into separate file/function, and only handle subjects updates here to improve maintainability and readability.